### PR TITLE
Add fixture-based e2e tests with sample apps

### PR DESCRIPTION
**Pull request by damemi**
Fixes https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/39

This adds an e2e test matrix (largely based on [collector-contrib's tests](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a45376db04289be6e0ca98276d26d8b504d9cad5/.github/workflows/e2e-tests.yml)) that does the following:

1. Builds the auto-instrumentation agent, launcher (to be removed with https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/40), and sample apps using each supported library
2. Spins up a kind cluster
3. Starts an otel collector with trace receiver and file exporter
4. Starts the sample app for that library
5. The sample app calls itself to generate trace data
6. Copies the file exporter output to a local fixture
7. Checks for changes to the current fixture, ignoring/removing timestamps

I used a k8s cluster instead of just running the binaries for a couple reasons: First, I wanted to just use a collector image from upstream rather than building it ourselves. Second, k8s gives some nice orchestration around starting up and stopping these processes that works well as a Github action. I couldn't find a good way to build local binaries and have those run in the background, while moving on to the next Workflow step, and checking that everything had completed successfully.

The steps are basically copied to a new `make fixture-$LIBRARY` target that lets you generate these fixtures. Eg, `make fixture-nethttp` will run all the steps from above. Failures in these tests could indicate a legitimate update to the fixtures, so updating them locally will be useful.

I added sample apps for gorillamux and net/http. If someone else wants to add a gRPC app, I think the first 2 serve as a good example. New libraries should be able to drop in similarly.

### CODE REVIEW

File: test/e2e/gorillamux/traces.json
Comment (damemi) [inline test/e2e/gorillamux/traces.json:45]: I was surprised that spanId and traceId were empty here. ping @edeNFed does that look right?

File: Makefile
Comment (MrAlias) [inline Makefile:71]: This will add a `fixture` target that shouldn't be called (`LIBRARY` will not be set if it is directly called). Instead, we could compose the built target

```suggestion
.PHONY: fixture-nethttp fixture-gorillamux
fixture-nethttp: fixtures/nethttp
fixture-gorillamux: fixtures/gorillamux
fixture/%: LIBRARY=$*
fixture/%:
```

File: .github/workflows/kind.yml
Comment (MrAlias) [inline .github/workflows/kind.yml:24]:
```suggestion
          go-version: "1.20"
```

File: test/e2e/gorillamux/Dockerfile
Comment (MrAlias) [inline test/e2e/gorillamux/Dockerfile:1]:
```suggestion
FROM golang:1.20
```

File: test/e2e/nethttp/Dockerfile
Comment (MrAlias) [inline test/e2e/nethttp/Dockerfile:1]:
```suggestion
FROM golang:1.20
```

File: test/e2e/gorillamux/traces.json
Comment (MrAlias) [inline test/e2e/gorillamux/traces.json:31]: It also seems weird that these values are empty (not "Get" for example).

**Review by pellared — COMMENTED** [review]
This is a comment rather for future improvment. I do not want to block this PR as we need to start with something 😉 

Based on my experience in https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation (where I am a maintainer) I find testing against a  "test fake collector" makes the tests more maintainable:

- spinning up a a `httptest` server or by setting up a gRPC server in tests is faster than running a real collector [image]
- you get more flexibility in assertion logic


Here you can find an example implementation and usage of a "test fake collector" in Go: https://github.com/signalfx/splunk-otel-go/blob/main/distro/otel_test.go

**Comment by damemi** [comment]
@pellared that's a great suggestion, but I agree on not blocking this PR right now (especially because I want to unblock #40). I'll make a new issue from your comment to track that for a refactor in the future.

Right now I just want to figure out the empty labels from above which I'm trying a couple things for. Maybe an issue with context propagation from the request?

File: test/e2e/gorillamux/traces.json
Comment (damemi) [reply test/e2e/gorillamux/traces.json:45]: Ah I didn't enable a propagator. Setting `OTEL_PROPAGATORS` on the sample app fills these in (and the request fields too).

File: test/e2e/gorillamux/traces.json
Comment (damemi) [reply test/e2e/gorillamux/traces.json:31]: https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/46#discussion_r1152142583

File: test/e2e/gorillamux/traces.json
Comment (damemi) [reply test/e2e/gorillamux/traces.json:73]: hm but this empty span is still here

**Comment by damemi** [comment]
It looks like https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/55 broke auto-instrumentation due to a change in [v0.9.1](https://github.com/cilium/ebpf/releases/tag/v0.9.1) that updated `UprobeOptions.Offset` to be relative, with `Address` now being absolute. Updated this in https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/46/commits/dddbb7e637537506c241d9988d43674c4019be51 and running e2es locally now seem to work again, hopefully they pass in CI too

**Comment by damemi** [comment]
They're green! 🎉 🎉

File: test/e2e/gorillamux/traces.json
Comment (damemi) [reply test/e2e/gorillamux/traces.json]: I think this empty span is a bug in the auto-instrumentation, but that should be a separate issue to track instead of fix in this PR

**Review by MrAlias — APPROVED** [review]

**Review by edeNFed — APPROVED** [review]

**Pull request merged**
